Pathways and levels

LeoFrachet opened the pull request:
Hi GTFS community!

As explained last October in the issue #108, and as discussed in the Google Docs linked in the issue ([bit.ly/gtfs-pathways](http://bit.ly/gtfs-pathways)), we've been working on extending GTFS to describe subway stations. Since the conversations faded out and since the consensus was reached, some agencies started to gather the data this Winter, including SF MTA, NYMTA & MBTA. They haven't put (as of today) those new fields in their production dataset, but we should be close to that for some of them.

From the full proposal, we extracted the mostly used fields, that you can either see in a Google Doc ([bit.ly/gtfs-pathways-core](http://bit.ly/gtfs-pathways-core)) or in this proposal.

A first dataset has been created for King's Cross. It's publicly available for those who want to see what it looks like, or who want to start implementing: http://bit.ly/gtfs-pathways-tfl-kings-cross (the PR and the Google Doc are slightly different, since I updated the phrasing of the definition form the old version "The `blabla` field defines a something that" into the new phrasing: "Something that")

gcamp commented:
I'm assuming stop_ids in stop_times.txt now can either be a Stop, Platform (type=0) or Boarding Area (type=4), but not Station (type=1), entrance (type=2) or generic nodes (type=3). Should that specified somewhere?

LeoFrachet commented:
Good point. I’ll add it.

Le ven. 15 févr. 2019 à 17:32, Guillaume Campagna <notifications@github.com>
a écrit :

> I'm assuming stop_ids in stop_times.txt now can either be a Stop, Platform
> (type=0) or Boarding Area (type=4), but not Station (type=1), entrance
> (type=2) or generic nodes (type=3). Should that specified somewhere?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/google/transit/pull/143#issuecomment-464229054>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ADSAVBB-5_EnNaaDsPIqyDk3Qn3uT5FVks5vNzWbgaJpZM4a-f2U>
> .
>
-- 
—
Leo Frachet | fr, en, de | MobilityData Technical Co-Director |
MobilityData.org
leo@MobilityData.org | CA & US: +1 514 222 7204 | FR & EU : ‭+33 7 69 87 85
13‬
Montréal, Québec, Canada

skinkie commented:
If boarding areas are would be routable, this proposal becomes even a bigger mess. The authors were against a much more simple extension to allow an entrance to a platform. If the proposal is extending GTFS with a transit routable boarding area but not directly make stations transit routable as well, it seems clear that the only intention is to make GTFS more complex, and not better usable.

LeoFrachet commented:
Guillaume (@gcamp): On second thought, no, I don't see what you're speaking about. StopTimes can only refer to stops (`location_type=0`). I agree that the definition in `stop_times.txt` should be updated from:

> Referenced locations must be stops, not stations or station entrances.

To:

> Referenced locations must be stops (`location_type=0`).

LeoFrachet commented:
Another dataset is publicly available. It contains the data for station Taverners Hill in Sydney (AU-NSW): [http://bit.ly/gtfs-pathways-tfnsw-taverners-hill](http://bit.ly/gtfs-pathways-tfnsw-taverners-hill).

Both datasets (King's Cross and Taverners Hill) are currently used in the production version of Google Maps. Both to process walking time in station and to display "follow sign" information (from `signposted_as` field):

<img width="1101" alt="capture d ecran le 2019-03-05 a 14 55 10" src="https://user-images.githubusercontent.com/3440724/53833555-d2ed1180-3f56-11e9-8e3d-331a442d987c.png">

LeoFrachet commented:
So we have data producers, which have made extract of their data public (for TfL & TfNSW stations) and we have a data consumer, Google Maps, using this data in production. Also the PR has been open for more than one week. 

Therefore I opening the vote on the PR. The vote will be open until Tuesday 12 March at 23:59:59 UTC.

skinkie commented:
@LeoFrachet is this a bogus claim? Previously it has been stated that Google was only using some parts of the pathway proposal, that previously existed and didn't have anything to do with your additions to it.

-1

LeoFrachet commented:
I'll let Google answer directly.

paulswartz commented on gtfs/spec/en/reference.md:
```diff
@@ -121,16 +125,17 @@ File: **Required**
 |  `stop_url` | URL | Optional | URL of a web page about the location. This should be different from the `agency.agency_url` and the `routes.route_url` field values. | 
-|  `location_type` | Enum | Optional | Indicates whether this `stop_id` represents a stop, station, or station entrance/exit (denoted as station entrance in this document). If `location_type` is omitted, all locations are treated as stops. Valid options are: <br><br>`0` or empty - Stop. A location where riders board or disembark from a transit vehicle.<br>`1` - Station. A physical structure or ar ...
-|  `parent_station` | ID referencing `stops.stop_id` | Optional | Identifies the station that a stop or station entrance is part of. This station-stop relationship, called a parent-child relationship, can only exist between a station (as the parent) and a stop or station entrance (as the child). These types are defined by location_type. If the stop or station identified in this record does not ha ...
+|  `location_type` | Enum | Optional | Type of the location:<br>• `0` (or blank): **Stop** (or **Platform**). A location where passengers board or disembark from a transit vehicle. Is called a platform when defined within a `parent_station`.<br>• `1`: **Station**. A physical structure or area that contains one or more platform.<br>• `2`: **Entrance/Exit**. A location where passengers can enter or ...
```
typo: "or but no"

paulswartz commented on gtfs/spec/en/reference.md:
```diff
@@ -121,16 +125,17 @@ File: **Required**
-|  `location_type` | Enum | Optional | Indicates whether this `stop_id` represents a stop, station, or station entrance/exit (denoted as station entrance in this document). If `location_type` is omitted, all locations are treated as stops. Valid options are: <br><br>`0` or empty - Stop. A location where riders board or disembark from a transit vehicle.<br>`1` - Station. A physical structure or ar ...
-|  `parent_station` | ID referencing `stops.stop_id` | Optional | Identifies the station that a stop or station entrance is part of. This station-stop relationship, called a parent-child relationship, can only exist between a station (as the parent) and a stop or station entrance (as the child). These types are defined by location_type. If the stop or station identified in this record does not ha ...
+|  `location_type` | Enum | Optional | Type of the location:<br>• `0` (or blank): **Stop** (or **Platform**). A location where passengers board or disembark from a transit vehicle. Is called a platform when defined within a `parent_station`.<br>• `1`: **Station**. A physical structure or area that contains one or more platform.<br>• `2`: **Entrance/Exit**. A location where passengers can enter or ...
+|  `parent_station` | ID referencing `stops.stop_id` | **Conditionally Required** | Defines hierarchy between the different locations defined in `stops.txt`. It contains the ID of the parent location, as followed:<br>• **Stop/platform** (`location_type=0`): the `parent_station` field contains the ID of a station.<br>• **Station** (`location_type=1`): this field must be empty.<br>• **Entrance/exit ...
```
May conflict with the previous line: is a parent_station required for entrances (location_type=2) or not?

paulswartz commented on gtfs/spec/en/reference.md:
```diff
@@ -121,16 +125,17 @@ File: **Required**
+|  `parent_station` | ID referencing `stops.stop_id` | **Conditionally Required** | Defines hierarchy between the different locations defined in `stops.txt`. It contains the ID of the parent location, as followed:<br>• **Stop/platform** (`location_type=0`): the `parent_station` field contains the ID of a station.<br>• **Station** (`location_type=1`): this field must be empty.<br>• **Entrance/exit ...
 |  `stop_timezone` | Timezone | Optional | Timezone of the location. If the location has a parent station, it inherits the parent station’s timezone instead of applying its own. Stations and parentless stops with empty `stop_timezone` inherit the timezone specified by `agency.agency_timezone`. If `stop_timezone` values are provided, the times in [stop_times.txt](#stop_timetxt) should be entered a ...
 |  `wheelchair_boarding` | Enum | Optional | Indicates whether wheelchair boardings are possible from the location. Valid options are: <br><br>For parentless stops:<br>`0` or empty - No accessibility information for the stop.<br>`1` - Some vehicles at this stop can be boarded by a rider in a wheelchair.<br>`2` - Wheelchair boarding is not possible at this stop. <br><br>For child stops: <br>`0` or ...
+|  `level_id` | ID referencing `levels.txt` | Optional | Level of the location. The same level can be used by multiple unlinked stations.| 
```
`levels.level_id` (not levels.txt, per some other table documentation)

aababilov commented:
Hello from Google,
We fully support Core GTFS-Pathways since 21 Feb 2019. This is demonstrated by screenshots posted by Léo.

+1

gcamp commented:
Can we have more details on what Google support and what is the "Core GTFS-Pathways" @aababilov ?

Also @LeoFrachet can you fix the conflicts before calling for a vote since there are going to be changes needed to be done anyway?

LeoFrachet commented:
The "Core GTFS-Pathways" is what is in this proposal. Which is what Google support. It is a subset of the full GTFS-Pathways which is in the document bit.ly/gtfs-pathways.

And yes I'll fix the conflicts.

dbabramov commented:
More details on Google's support for the proposed specification:

1. All new fields have been added to Google's validation pipelines, new warnings and errors (such as invalid values, unreachable entities, trips serving entities non-stops) have been added.

2. pathways.txt
Google used to support single-step entrance-to-platform pathways in the past.
Support has been added for multi-step pathway routing between station entities (pairs of platforms or entrances-to-platforms) via generic nodes.
New fields:
* is_bidirectional: support for bi-directional pathways has been added;
* pathway_mode: not exposed directly in the UI; used in the evaluations of walking times between station entities and penalisation of walking routes that lead through multiple paygates;
* traversal_time: used in calculation of walking times;
* stair_count: does not show up directly in the UI, used in calculation of walking times;
* signposted_as, reverse_signposted_as: shown in the UI
* min_width and max_slope: not used at the moment, reserved for future accessibility support

3. Changes to stops.txt:
* Location types 3 (generic node) and 4 (boarding area) are used in multi-step routing between station entities (entrance-to-platform and platform-to-platform), and in calculation of walking times.

4. Changes to levels.txt:
* level_id, level_index, level_name - will not appear with routing results immediately. Used to link station entities to indoor modelling of stations where available.

abyrd commented:
I am now trying to consolidate information from throughout this thread so I can formulate an opinion on it. 

@gcamp my understanding is that "core pathways" means the set of features listed in the document http://bit.ly/gtfs-pathways-core, and the changes in this PR are intended to represent exactly that set of features, which are a subset of the larger recent pathways proposal. However "core pathways" is a superset of the older pathways proposal from years in the past.

@aababilov and @dbabramov have confirmed that Google has implemented the entire set of features listed in the above-linked document, and would display or otherwise use all new data described in this pull request. 

One source of concern is that this micro-mapping is relatively complex and would probably require special tooling to be widely adopted. Some less complex changes focusing on implicit connectivity among grouped entities (rather than explicit pathway topology) could promote better representation of compound stations across a larger number of data sets. But perhaps that should be a separate proposal.

There are also some decisions mixed into this PR (and the above comments) that might be considered orthogonal to the core issue of describing station pathway topology, specifically the idea that stop_times can only reference stops, meaning trips can only stop at specific platforms, rather than stations (when the platform is dynamically allocated and not yet known).

abyrd commented:
@LeoFrachet shall I create an issue for stop_times visiting stations, or were you already planning to create one including the information you've gathered?

On the topic of this PR, I am wondering about the inclusion of the additional levels.txt table in the minimal "core" proposal. I see that the levels section is a subset of the original at
https://docs.google.com/document/d/1qJOTe4m_a4dcJnvXYt4smYj4QQ1ejZ8CvLBYzDM5IyM/edit#bookmark=kix.tn25emxn4pie. But in reading that document and the core pathways doc at https://docs.google.com/document/d/1gQZBHrgBYnfSBDTeIeE3a8YMs7RKYaFjdNQm1Lz7sPU/edit I still don't fully understand the purpose. They say a levels table is required to tell people to get off at named floors, but a separate levels table is only one way to provide that information. The idea of defining feed-wide building levels that can be reused in different stations seems more strictly normalized than other parts of GTFS. It's also debatable whether a mezzanine level +1 in station A is really the same entity as a mezzanine level +1 in station B, as they are not the same physical floor (and not mutually connected to one another). I also don't see how such unification of levels would improve trip planning.

Might it be just as effective, yet less complex (avoiding a join), to include the level_index and level_name fields in the stops themselves? By way of comparison, we don't have a `headsigns.txt` table where many trips can reference the same headsign string.

LeoFrachet commented:
Regarding scheduling on station, I opened an issue: https://github.com/google/transit/issues/147

Regarding `levels.txt` vs just adding fields in `stops.txt`, I have no strong opinion (in fact I don't even have an opinion at all). I do not see one as being better than the other one. This is a part of the spec that I just kept from the original proposal.

But my concern is that now, in March 2019, we already have quite a few implementations done(MBTA, NY MTA, SF MTA, Google, London and some others stakeholders working on it). I don't know if all of them are supporting `levels.txt` (some might have no elevator, even if I doubt it). But it might be a bit late to change it if the two options are more or less equivalent and that `levels.txt` isn't creating big issue.

paulswartz commented:
> But my concern is that now, in March 2019, we already have quite a few implementations done(MBTA, NY MTA, SF MTA, Google, London and some others stakeholders working on it). I don't know if all of them are supporting `levels.txt` (some might have no elevator, even if I doubt it). But it might be a bit late to change it if the two options are more or less equivalent and that `levels.txt` isn't creating big issue.

@mbta has a `levels.txt` (42 rows) in our public feed.

LeoFrachet commented:
I've resolved the conflicts (ping Guillaume @gcamp) and I've fixed the requirements of parent station for entrances which was contradictory (thanks Paul @paulswartz). I changed it to always required (in `location_type` it was said to be optional if it was belonging to multiple stations, which was inconsistent with the definition of `parent_station` which was required for entrance). I also fixed typos that Paul pointed out.

So the PR changed.

While this, the first vote didn't pass (only one +1 by Google and one -1 by Stefan, so both not-enough voters and a veto, which make sense since some questions were still open).

Since the PR changed, I'm rebooting the vote. **The vote will be open until Thursday 19 March at 23:59:59 UTC.**

Stefan (@skinkie) and Dmitri (@dbabramov) please revote.
Paul (@paulswartz), Pascal (@prhod), Guillaume (@gcamp) and Andrew (@abyrd), don't hesitate to vote if you want to.

aababilov commented:
+1

skinkie commented:
+1

dbabramov commented:
+1

kurtraschke commented:
+1

prhod commented:
+1

paulswartz commented:
+1, though it would be nice to clean up the trailing whitespace before merging.

abyrd commented:
+0. I am hesitant to block this change alone at the last minute, as I see that a large number of stakeholders are already structuring the data this way with a separate levels table. It would be helpful though if someone can explain why levels are broken out into a separate table. I do not see the two representations (embedded field versus reference to external table) as equivalent. The table option somehow implies that the mezzanine or basement level of two unconnected stations are the same entity. It would be preferable to make such a decision about the data model for a solid reason. 

Fields stating the level directly (rather than referencing external tables) seem more readable, not significantly more voluminous, and closer to the reality of different stations being different entities.

LeoFrachet commented:
So the vote passed!

5 votes in favor:
- Google: Alexej @aababilov & Dmitri @dbabramov 
- Stefan @skinkie
- NY MTA: Kurt @kurtraschke
- Kisio: Pascal @prhod
- MBTA: Paul @paulswartz

One abstention:
- Conveyal: Andrew @abyrd

I'll fix the trailing spaces and the CLA and I'll merge.

LeoFrachet commented:
Trailing spaces removed.

Alexej @aababilov can you force-approve the CLA? Thanks!

aababilov commented:
cla: yes

mmorang commented:
Can someone post a link to a complete GTFS dataset that makes use of these new tables?  The examples linked above for Taverner's Hill and King's Cross have only a subset of the entire GTFS feed.  I need a full dataset to do some testing for my tools.

sdjacobs commented on gtfs/spec/en/reference.md:
```diff
@@ -312,6 +317,47 @@ When calculating an itinerary, GTFS-consuming applications interpolate transfers
+|  Field Name | Type | Required | Description |
+|  ------ | ------ | ------ | ------ |
+|  `level_id` | ID | **Required** | Id of the level that can be referenced from `stops.txt`.|
+|  `level_index` | Float | **Required** | Numeric index of the level that indicates relative position of this level in relation to other levels (levels with higher indices are assumed to be located above levels with lower indices).<br><br>Ground level should have index 0, with levels above ground indicated by positive indices and levels below ground by negative indices.|
```
`level_index` is an integer in the spec [here](http://bit.ly/gtfs-pathways-core), rather than a float. Which is correct?

paulswartz replied:
@mbta is treating it as a float, and we have some non-integer values already.
